sdk/go/pulumi/rpc_test.go

```diff
@@ -829,3 +830,56 @@ func TestInvalidArchive(t *testing.T) {
 	_, _, err = marshalInput(d, archiveType, true)
 	assert.Error(t, err)
 }
+
+func TestDependsOnComponent(t *testing.T) {
```

Comment:
Maybe this is our chance to be just a little more verbose and record the reference knowledge here, for example I'm still a little confused about the different treatment for e.g. CustomResource, we could add a test to make it obvious it's intentional, and add a few comments here. Alternatively if we have a good doc reference for intended behavior, we could link it here.